source/validate_type_unique.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Ensures type declarations are unique unless allowed by the specification.
+
+#include "validate.h"
+
+#include "diagnostic.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+// Validates that type declarations are unique, unless multiple declarations
+// of the same data type are allowed by the specification.
+// (see section 2.8 Types and Variables)
+spv_result_t TypeUniquePass(ValidationState_t& _,
+                            const spv_parsed_instruction_t* inst) {
+  const auto opcode = inst->opcode;
+  if (opcode == SpvOpTypeArray || opcode == SpvOpTypeStruct) {
+    // Duplicate declarations of aggregates are allowed.
+    return SPV_SUCCESS;
+  }
+
+  if ((opcode >= SpvOpTypeVoid && opcode <= SpvOpTypeForwardPointer) ||
```

Comment:
This expression captures the condition that the opcode is a type-creating instruction.   It seems a bit fragile. 
It's better to centralize it.

I just found spvOpcodeGeneratesType in opcode.h.  That would be better to use.
However, we need to fix it to add PipeStorage and NamedBarrier.
Also, it knowingly excludes ForwardPointer.  OpTypeForwardPointer does not create a new type, so it should be excluded from the uniqueness rule, I think.